### PR TITLE
value class codec macros

### PR DESCRIPTION
**Pull request by kubinio123**

### CODE REVIEW

File: core/src/main/scala-2/sttp/tapir/internal/CodecValueClassMacro.scala [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    val tpe: Type = weakTypeOf[T]
+
+    val isValueClass: Boolean = {
+      import definitions._
+      tpe.typeSymbol.isClass && tpe <:< AnyValTpe && !ScalaPrimitiveValueClasses.contains(tpe.typeSymbol.asClass)
+    }
+
+    val constructor = tpe.decls.collectFirst { case m: MethodSymbol if m.isPrimaryConstructor => m }
+
+    val fields = constructor.get.paramLists.head
```
Comment (mbore) [inline core/src/main/scala-2/sttp/tapir/internal/CodecValueClassMacro.scala:22]: I think that `fields` and `constructor` should be `lazy` to not process them in non value classes cases.

File: core/src/main/scala-3/sttp/tapir/internal/CodecValueClassMacro.scala [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    def encodeBody[F: Type](term: Term): Expr[F] = Select(term, field).asExprOf[F]
+
+    def encode[F: Type]: Expr[T => F] = '{ (t: T) => ${ encodeBody[F]('{ t }.asTerm) } }
+
+    fieldTpe.asType match
+      case '[f] =>
+        val dec = decode[f]
+        val enc = encode[f]
+
+        '{ $baseCodec.asInstanceOf[Codec[String, f, TextPlain]].map($dec)($enc) }
```
Comment (mbore) [inline core/src/main/scala-3/sttp/tapir/internal/CodecValueClassMacro.scala:46]:
```suggestion
    fieldTpe.asType match
      case '[f] =>
        val baseCodec = Expr.summon[Codec[String, f, TextPlain]].getOrElse {
          report.errorAndAbort(
            s"Cannot summon codec for value class ${tpe.show} wrapping ${fieldTpe.show}."
          )
        }

        val dec = decode[f]
        val enc = encode[f]

        '{ $baseCodec.map($dec)($enc) }

```

File: core/src/main/scala-2/sttp/tapir/internal/CodecValueClassMacro.scala (hunk 1 above)
Comment (kubinio123) [reply core/src/main/scala-2/sttp/tapir/internal/CodecValueClassMacro.scala:22]: i will put them in `else` branch then

**Comment by kubinio123** [comment]
thanks @mbore ;)

**Comment by adamw** [comment]
(maybe as a separate task)

**Comment by kubinio123** [comment]
Doesn't `schemaForCaseClass` handle that?

**Comment by adamw** [comment]
@kubinio123 Ah right, this is treated as a product by magnolia :)

**Pull request merged**